packages/app/src/app/location-selection/polygon-map.service.ts

```diff
@@ -0,0 +1,430 @@
+  }
+
+  /**
+   * Fetch polygons from API and render them on the map
+   * @param projectId project ID to filter polygons by
+   */
+  private fetchAndRenderPolygons(projectId: number): void {
+    if (!this.map) {
+      console.warn('Map not initialized, cannot fetch polygons');
+      return;
+    }
+
+    this.loading.set(true);
+
+    this.api
+      .getPolygons({ projectId })
+      .pipe(
+        tap(response => {
+          console.debug(`Fetched polygons for project ${projectId}`, response.polygons);
+          this.polygons.set(response.polygons);
+          this.polygons$.next(response.polygons);
+        }),
+        finalize(() => this.loading.set(false)),
+        takeUntilDestroyed(this.destroyRef)
+      )
+      .subscribe({
+        next: response => {
+          this.renderPolygonsOnMap(response.polygons, projectId);
+        },
+        error: error => {
+          console.error('Error fetching polygons:', error);
+          this.snackbar.open('Failed to load polygons', 'OK', {
+            duration: 3000
+          });
+        }
+      });
+  }
+
+  /**
+   * Intelligently update polygon layers on the map.
+   * Only adds/updates/removes polygons that have changed.
+   * @param polygons Array of polygon data from API
+   * @param projectId Optional project ID for layer naming
+   */
+  private renderPolygonsOnMap(polygons: PolygonReference[], projectId: number): void {
+    if (!this.map) {
+      console.warn('Map not initialized');
+      return;
+    }
+
+    // Setup layer group if needed
+    const layerGroup = this.setupPolygonLayerGroup(projectId);
+
+    // Get the vector layer (create if doesn't exist)
+    let vectorLayer = this.getOrCreateVectorLayer(layerGroup);
+    const source = vectorLayer.getSource();
+
+    if (!source) {
+      console.error('Vector layer has no source');
+      return;
+    }
+
+    // Build set of new polygon IDs
+    const newPolygonIds = new Set(polygons.map(p => p.id));
+
+    // Track current polygon IDs
+    const currentPolygonIds = new Set(this.polygonsById.keys());
+
+    // 1. Remove polygons that no longer exist
+    for (const polygonId of currentPolygonIds) {
+      if (!newPolygonIds.has(polygonId)) {
+        console.debug(`Removing polygon ${polygonId} from map`);
+        const feature = this.featuresByPolygonId.get(polygonId);
+        if (feature) {
+          source.removeFeature(feature);
+          this.featuresByPolygonId.delete(polygonId);
+        }
+        this.polygonsById.delete(polygonId);
+      }
+    }
+
+    // 2. Add new polygons or update changed ones
+    for (const polygon of polygons) {
+      const existingPolygon = this.polygonsById.get(polygon.id);
+
+      // Check if polygon is new or has changed
+      const isNew = !existingPolygon;
+      const hasChanged = existingPolygon && this.hasPolygonChanged(existingPolygon, polygon);
+
+      if (isNew) {
+        console.debug(`Adding new polygon ${polygon.id} to map`);
```

Comment:
do we want all these logs on? could consider some kind of logging lib if we want logging configurability for developers